core/app/services/spree/cart/add_item.rb

```diff
@@ -32,6 +32,8 @@ def add_to_line_item(order:, variant:, quantity: nil, options: {})
         end
 
         line_item.target_shipment = options[:shipment] if options.key? :shipment
+        line_item.public_metadata = public_metadata.to_h if public_metadata
```

Comment:
@aleksandarpetrushev why do we need to call `.to_h` here?

Reply:
Because it does not implicitly cast the param to hash for some reason, but I have moved this casting to the controller.

Reply:
@aleksandarpetrushev if it's needed, than let's do it, but not in the controller, service is the right place for such logic

Reply:
Will do. Without casting it just does not assign the value to the attribute of the model.

Reply:
@aleksandarpetrushev so how will this behave with an invalid value? will this resolve in 50x error (not acceptable) or 422?

Reply:
we need a test for such scenario

Reply:
it does not fail at all, just discards the invalid value. Because when permitting the metadata params it expects a hash, so if it is for example a string, the permitted param will be nil. Should we fail explicitly if we have invalid metadata params?

Reply:
@aleksandarpetrushev yes, we should fail, otherwise it's hard to debug for a developer using an api